File: problem169.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        last = nums[0]
        count=1
        
        for i in range(1, len(nums)):
            if nums[i]==last:
                count+=1
            else:
                if count>0:
                    count=count-1
                else:
                    last=nums[i]
                    count=1
                
        return last
        
        
#time: O(N)
#space: O(1)
# https://www.geeksforgeeks.org/boyer-moore-majority-voting-algorithm/

File: problem169.cs
/*

[2,2,1,1,1,2,2]


*/


public class Solution {
    public int MajorityElement(int[] nums) {
        int last = nums[0];
        int count=1;
        
        for(int i=1; i<nums.Length; i++)
        {
            if(nums[i] == last)
            {
                count++;
            }
            else if(count>0)
            {
                count--;
            }
            else
            {
                last=nums[i];
                count=1;
            }
        }
        return last;
        
    }
}

//time: O(N)
//space: O(1)